text/0000-tracked-properties.md

```diff
@@ -0,0 +1,1029 @@
+
+class Person {
+  @tracked firstName = 'Tom';
+  @tracked lastName = 'Dale';
+
+  @tracked
+  get fullName() {
+    return `${this.firstName} ${this.lastName}`;
+  }
+}
+```
+
+### Getting Tracked Properties
+
+Tracked properties can be accessed using standard Javascript syntax. From the
+user's point of view, there is nothing special about them. This should continue
+to work in the future, even if new methods are added for accessing properties,
+because tracked properties use native getters under the hood.
+
+```js
+let person = new Person();
+
+// Dot notation
+const fullName = person.fullName;
+// Destructuring
+const { fullName } = person;
+// Bracket notation for computed property names
+const fullName = person['fullName'];
+```
+
+### Setting Tracked Properties
+
+Tracked properties can be set using standard Javascript syntax. They use native
+setters under the hood, meaning that there is no need for using a setter method
+like `set`.
+
+```js
+let person = new Person();
+
+// Simple assignment
+person.firstName = 'Jen';
+// Addition assignment (+=)
+person.lastName += 'Weber';
+// Increment operator
+person.age++;
+```
+
+### Autotracking
+
+Tracked properties do not need to specify their dependencies. Under the hood,
+this works by utilizing an _autotrack stack_. This stack is a bit of global
+state which tracked getters can access. As tracked getters and properties are
+accessed, they push themselves onto the stack, and once they have finished
+running, the stack contains the full list of all the tracked properties that
+were accessed while it was running.
+
+In our first example, with the `Person` class, we can see this in action:
+
+```js
+import { tracked } from '@glimmer/tracking';
+
+class Person {
+  @tracked firstName = 'Tom';
+  @tracked lastName = 'Dale';
+
+  @tracked
+  get fullName() {
+    return `${this.firstName} ${this.lastName}`;
+  }
+}
+
+let person = new Person();
+```
+
+When we create a new instance of `Person`, `fullName` has no knowledge of
+`firstName` or `lastName`. If we set either of those values now, it won't know
+that anything has changed:
+
+```js
+// Dirties the `firstName` property, but because `fullName` has not been
+// accessed yet nothing happens to it.
+person.firstName = 'Rob';
+```
+
+This is ok though, because _nothing_ has accessed `fullName` yet. There is no
+state to invalidate anywhere else. Now, let's say we access `fullName`, and then
+update the field again:
+
+```js
+person.fullName; // 'Rob Dale'
+
+person.lastName = 'Jackson';
+```
+
+Now `fullName` knows which tracked properties were accessed when it was run,
+and setting `lastName` has invalidated `fullName`.
+
+> **NOTE:** This does _not_ invalidate a cache like in computed properties. Even
+> if `firstName` and `lastName` were untracked, the tracked getter would still
+> return the correct value on subsequent accesses, because `@tracked` does _not_
+> cache values. The validation and invalidation is pure meta data that is only
```

Comment:
where is the metadata stored?

what's the reasoning behind not storing that metadata on a private class field?

Reply:
The metadata is currently stored in the object's `meta`, though this will need to be generalized to work with both Ember.js and Glimmer.js. This info needs to be readable by things _outside_ of the class, so private fields wouldn't really work here. Also, I'm pretty sure private class fields have to be declared at class definition time.